Test: Added tcpdump containers on test and dump output when it fails.

eloycoto opened the pull request:
This commit try to get visibility when the cilium restarts on Kubernetes
test. Tcpdump containers will be in the backend running on host
networking fetching all the traffic and will be dump in logs if a test
failed.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5843)
<!-- Reviewable:end -->

eloycoto committed:
Test: Added tcpdump containers on test and dump output when it fails.

This commit try to get visibility when the cilium restarts on Kubernetes
test. Tcpdump containers will be in the background running on pods in
all host and all the traces will be saved in there.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

aanm commented on test/k8sT/manifests/tcpdump.yaml:
```diff
@@ -0,0 +1,47 @@
+        prometheus.io/port: "9090"
+    spec:
+      containers:
+        - image: corfr/tcpdump
```
add image tag, even if it's latest

aanm commented on test/helpers/kubectl.go:
```diff
@@ -396,6 +396,61 @@ func (kub *Kubectl) MicroscopeStart() (error, func() error) {
+			return err
+		}
+
+		for pod, response := range tcpdumpCmdsResponses {
```
I know what you're doing here, this can take a lot of time, we can be smart and write the output of tcpdump directly to a file with the `-w` flag. The daemonset would mount a directory in the VM where it's running on. Instead getting the container logs we would simply copy the file already written.

eloycoto replied:
can't do that because the report folder is not in there until AfterEach is in there.

aanm replied:
I was thinking something like this:

1. `tcpdump -i any ... -w /directory/mounted/in/pod/foo.pcap`
..
2. testPath, err := CreateReportDirectory()
..
3. response.WaitUntilFinish()
4. kubectl.Exec("cp /directory/mounted/in/pod/foo.pcap /report/directory")

I don't know how feasible 4 is but we do gather `kubelet` logs of all nodes so we can do the same thing we do with kubelet.

eloycoto replied:
I do not really liked: 

- can't run the tcpdump in other vms other than ours (minukube)
- need to create the folder before, not a issue,at all. 
- per each tcpdump needs to be added a prefix with the hostname to save it. 

I'll change it tomorrow/monday.

aanm replied:
>can't run the tcpdump in other vms other than ours (minukube)

how is that a problem? `tcpdump` is running inside the pod 

> need to create the folder before, not a issue,at all.

it's created by the daemonset automatically

>per each tcpdump needs to be added a prefix with the hostname to save it.

it can be available as an env variable

eloycoto replied:
> how is that a problem? tcpdump is running inside the pod

The /vagrant/ or /go/src/github.com/cilium/cilium/ is not shared as we have in the test VM. 

> it's created by the daemonset automatically

No, the test report folder is not shared, copy files from one side to another does not make sense to me at all. 

> it can be available as an env variable

yes, but for copy and paste, need to get that env. 

It's ok, I'll do the changes, tomorrow or Monday. 

Best regards

aanm replied:
> > how is that a problem? tcpdump is running inside the pod
> 
> The /vagrant/ or /go/src/github.com/cilium/cilium/ is not shared as we have in the test VM.
> 
What does that have to do with `tcpdump` execution?
> > it's created by the daemonset automatically
> 
> No, the test report folder is not shared, copy files from one side to another does not make sense to me at all.
> 
can you clarify what side to another are you referring to?
> > it can be available as an env variable
> 
> yes, but for copy and paste, need to get that env.
> 
What copy paste are you referring to?
> It's ok, I'll do the changes, tomorrow or Monday.
> 
> Best regards

eloycoto replied:
André, don't worry, I'll add the changes, all make sense, please don't spend more time on this one. 
Thanks

ianvernon commented on test/k8sT/Services.go:
```diff
@@ -79,11 +81,14 @@ var _ = Describe("K8sServicesTest", func() {
 	JustAfterEach(func() {
 		kubectl.ValidateNoErrorsOnLogs(CurrentGinkgoTestDescription().Duration)
 		Expect(microscopeCancel()).To(BeNil(), "cannot stop microscope")
+		Expect(tcpdumpCancel()).To(BeNil(), "cannot stoped tcpdump")
```
stoped --> stop

ianvernon commented on test/k8sT/Updates.go:
```diff
@@ -62,6 +65,12 @@ var _ = Describe("K8sUpdates", func() {
 
 	JustAfterEach(func() {
 		kubectl.ValidateNoErrorsOnLogs(CurrentGinkgoTestDescription().Duration)
+		Expect(tcpdumpCancel()).To(BeNil(), "cannot stoped tcpdump")
```
stoped --> stop

ianvernon commented on test/test_suite_test.go:
```diff
@@ -312,5 +313,14 @@ var _ = AfterEach(func() {
 		_ = os.Remove(filepath.Join(testPath, helpers.MonitorLogFileName))
+
+		// Removed all tcpdump files
+		tcpdumpFiles, err := filepath.Glob(fmt.Sprintf("%s/%s-*.log", testPath, helpers.Tcpdump))
```
What is the need to do this in the global after-each?

eloycoto replied:
tcpdump files are big, so if the test didn't fail we deleted the files to not store it.

ianvernon replied:
Got it, thanks - can you update the comment to indicate that?

eloycoto committed:
Test: Tcpdump using a Kubernetes volume.

Requested by André on first review. To store all in a Volume and save
all the data directly in the folder instead of dumping the log using
ssh.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

aanm commented on test/k8sT/manifests/tcpdump.yaml:
```diff
@@ -0,0 +1,55 @@
+      volumes:
+        - name: test-volume
+          hostPath:
+            path: /home/vagrant/go/src/github.com/cilium/cilium/test/
```
can't this be `/tmp` to work in multiple environments?

eloycoto replied:
Spoke offline with andré, instead of using volumes, we'll get the file using `kubectl cp`

aanm commented on test/helpers/kubectl.go:
```diff
@@ -396,6 +396,56 @@ func (kub *Kubectl) MicroscopeStart() (error, func() error) {
+	var containerVolumeMount = "/test/"
+	var cb = func() error { return nil }
+
+	testPath, err := CreateReportDirectory()
```
`s/testPath/reportDir/` or `s/testPath/reportPath/`

aanm commented on test/helpers/kubectl.go:
```diff
@@ -396,6 +396,56 @@ func (kub *Kubectl) MicroscopeStart() (error, func() error) {
+		return err, cb
+	}
+
+	_ = kub.Apply(ManifestGet(TcpdumpManifest))
```
Why don't we check for errors here?

eloycoto replied:
because Tcpdump pod is never deleted. we only need to make sure that is in place.

eloycoto committed:
Test: Tcpdump use kubectl cp instead of volumes.

Added kubectl cp and do not use volumes.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>
